File: src/test/java/ru/jerael/booktracker/backend/application/usecase/book/CreateBookUseCaseImplTest.java
package ru.jerael.booktracker.backend.application.usecase.book;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import ru.jerael.booktracker.backend.domain.exception.NotFoundException;
import ru.jerael.booktracker.backend.domain.model.Genre;
import ru.jerael.booktracker.backend.domain.model.book.Book;
import ru.jerael.booktracker.backend.domain.model.book.BookCreation;
import ru.jerael.booktracker.backend.domain.model.book.BookStatus;
import ru.jerael.booktracker.backend.domain.repository.BookRepository;
import ru.jerael.booktracker.backend.domain.repository.GenreRepository;
import java.time.Instant;
import java.util.List;
import java.util.Set;
import java.util.UUID;
import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class CreateBookUseCaseImplTest {

    @Mock
    private BookRepository bookRepository;

    @Mock
    private GenreRepository genreRepository;

    @InjectMocks
    private CreateBookUseCaseImpl useCase;

    private final UUID userId = UUID.fromString("2c5781ea-1bc2-4561-a83d-26106df2526e");

    @Test
    void execute_WhenAllGenresFound_ShouldCreateBookWithAllGenres() {
        UUID id = UUID.fromString("ee39af7a-a073-4473-878a-1aae34e98bb7");

        String title = "title";
        BookStatus status = BookStatus.WANT_TO_READ;
        Set<Integer> genreIds = Set.of(1, 2);
        BookCreation data = new BookCreation(
            userId,
            title,
            status,
            genreIds,
            Set.of(),
            null,
            null,
            null,
            null,
            null,
            null,
            null
        );

        Genre genre1 = new Genre(1, "action");
        Genre genre2 = new Genre(2, "adventure");
        Set<Genre> genres = Set.of(genre1, genre2);

        when(genreRepository.findAllById(genreIds)).thenReturn(genres);
        when(bookRepository.save(any(), eq(userId))).thenAnswer(invocationOnMock -> {
            Book book = invocationOnMock.getArgument(0);
            return new Book(
                id,
                userId,
                book.title(),
                book.coverFileName(),
                book.createdAt(),
                book.genres(),
                Set.of(),
                null,
                null,
                null,
                null,
                null,
                null,
                null,
                List.of(),
                List.of()
            );
        });

        Book result = useCase.execute(data);

        assertNotNull(result);
        assertEquals(id, result.id());
        assertNotNull(result.createdAt());
        assertEquals(status, result.status());
        assertEquals(genres, result.genres());

        verify(genreRepository).findAllById(genreIds);

        ArgumentCaptor<Book> bookArgumentCaptor = ArgumentCaptor.forClass(Book.class);
        verify(bookRepository).save(bookArgumentCaptor.capture(), eq(userId));

        Book capturedBook = bookArgumentCaptor.getValue();
        assertNull(capturedBook.id());
        assertTrue(capturedBook.createdAt().isBefore(Instant.now().plusSeconds(2)));
    }

    @Test
    void execute_WhenOneOrMoreGenresNotFound_ShouldThrowNotFoundException() {
        String title = "title";
        Set<Integer> genreIds = Set.of(1, 2);
        BookCreation data = new BookCreation(
            userId,
            title,
            null,
            genreIds,
            Set.of(),
            null,
            null,
            null,
            null,
            null,
            null,
            null
        );

        Genre genre1 = new Genre(1, "action");
        when(genreRepository.findAllById(genreIds)).thenReturn(Set.of(genre1));

        assertThrows(NotFoundException.class, () -> useCase.execute(data));
        verify(bookRepository, never()).save(any(), eq(userId));
    }
}

File: src/test/kotlin/ru/jerael/booktracker/backend/application/usecase/book/CreateBookUseCaseImplTest.kt
package ru.jerael.booktracker.backend.application.usecase.book

import io.mockk.every
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.slot
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertNotNull
import org.junit.jupiter.api.assertNull
import org.junit.jupiter.api.extension.ExtendWith
import ru.jerael.booktracker.backend.domain.exception.NotFoundException
import ru.jerael.booktracker.backend.domain.model.Genre
import ru.jerael.booktracker.backend.domain.model.author.Author
import ru.jerael.booktracker.backend.domain.model.book.Book
import ru.jerael.booktracker.backend.domain.model.language.Language
import ru.jerael.booktracker.backend.domain.model.publisher.Publisher
import ru.jerael.booktracker.backend.domain.repository.*
import ru.jerael.booktracker.backend.domain.validator.BookValidator
import ru.jerael.booktracker.backend.factory.book.BookDomainFactory
import java.time.Instant
import java.util.*

@ExtendWith(MockKExtension::class)
class CreateBookUseCaseImplTest {
    
    @MockK
    private lateinit var bookRepository: BookRepository
    
    @MockK
    private lateinit var genreRepository: GenreRepository
    
    @MockK
    private lateinit var authorRepository: AuthorRepository
    
    @MockK
    private lateinit var publisherRepository: PublisherRepository
    
    @MockK
    private lateinit var languageRepository: LanguageRepository
    
    @MockK(relaxed = true)
    private lateinit var bookValidator: BookValidator
    
    @InjectMockKs
    private lateinit var useCase: CreateBookUseCaseImpl
    
    @BeforeEach
    fun setUp() {
        every { authorRepository.findByFullName(any()) } returns Optional.empty()
        every { authorRepository.save(any()) } answers { it.invocation.args[0] as Author }
        
        every { publisherRepository.findByName(any()) } returns Optional.empty()
        every { publisherRepository.save(any()) } answers { it.invocation.args[0] as Publisher }
        
        every { languageRepository.findByCode("en") } returns Optional.of(Language("en", "English"))
    }
    
    @Test
    fun `when all genres found, execute should create book with all genres`() {
        val genreIds = setOf(1, 2)
        val foundGenres = setOf(
            Genre(1, "Genre 1"),
            Genre(2, "Genre 2")
        )
        val userId = UUID.randomUUID()
        val bookId = UUID.randomUUID()
        val bookCreation = BookDomainFactory.createBookCreation(
            userId = userId,
            genreIds = genreIds,
            languageCode = "en"
        )
        
        every { genreRepository.findAllById(genreIds) } returns foundGenres
        every { bookRepository.save(any(), eq(userId)) } answers {
            val book = it.invocation.args[0] as Book
            BookDomainFactory.createBook(
                id = bookId,
                userId = userId,
                genres = book.genres
            )
        }
        
        with(useCase.execute(bookCreation)) {
            assertEquals(bookId, id)
            assertNotNull(createdAt)
            assertEquals(bookCreation.status, status())
            assertEquals(genres, this.genres)
        }
        
        verify { genreRepository.findAllById(genreIds) }
        
        val bookSlot = slot<Book>()
        verify { bookRepository.save(capture(bookSlot), eq(userId)) }
        
        with(bookSlot.captured) {
            assertNull(id)
            assertThat(createdAt).isBefore(Instant.now().plusSeconds(2))
        }
    }
    
    @Test
    fun `when one or more genres not found, execute should throw NotFoundException`() {
        val genreIds = setOf(1, 2)
        val foundGenres = setOf(Genre(1, "Genre 1"))
        val bookCreation = BookDomainFactory.createBookCreation(genreIds = genreIds, languageCode = "en")
        
        every { genreRepository.findAllById(genreIds) } returns foundGenres
        
        assertThrows(NotFoundException::class.java) { useCase.execute(bookCreation) }
        
        verify(exactly = 0) { bookRepository.save(any(), eq(bookCreation.userId)) }
    }
}